_posts/contests/2021-02-05-Performance-analysis-and-tuning-contest-4.md

```diff
@@ -96,4 +96,188 @@ __P.S.__ I'm also open to your comments and suggestions. Especially if you have
+  }
+}
+```
+
+The access pattern for the arrays in the innermost loop is good, it is always the sequential pattern. Everytime the loop counter `cc` in the innermost loop increases by one, the index into the `image` and `kernel` increases by one.
+
+The problem is the trip count of the innermost loop: it is low. Vectorization doesn't pay off in that case.
+
+One of the proposed solution was to multiversion a loop. In our case, `center` had value 2. We can create two versions of the loop nest, one where the value of `center` is known at compile time, and another one where it isn't. The compilers are very efficient at unrolling the loop with compile-time known trip count. Here is the solution:
+
+```cpp
+if (center == 2) {
+  for(r=0;r<rows;r++){
+    for(c=0;c<cols;c++){
+        dot = 0.0;
+        sum = 0.0;
+        for(cc=(-2);cc<=2;cc++){
+          if(((c+cc) >= 0) && ((c+cc) < cols)){
+              dot += (float)image[r*cols+(c+cc)] * kernel[center+cc];
+              sum += kernel[center+cc];
+          }
+        }
+        tempim[r*cols+c] = dot/sum;
+    }
+  }
+} else {
+  for(r=0;r<rows;r++){
+    for(c=0;c<cols;c++){
+        dot = 0.0;
+        sum = 0.0;
+        for(cc=(-center);cc<=center;cc++){
+          if(((c+cc) >= 0) && ((c+cc) < cols)){
+              dot += (float)image[r*cols+(c+cc)] * kernel[center+cc];
+              sum += kernel[center+cc];
+          }
+        }
+        tempim[r*cols+c] = dot/sum;
+    }
+  }
+}
+```
+
+The above approach is quite crude, we could have used C macros to achieve similar things with less copying. BUt the basic idea is there.
+
+Another approach is to do the loop interchange. If we could exchange the loop over `cc` and loop over `c`, we would get the innermost loop with a high trip count. Loop interchange is possible if two loops are perfectly nested, which is not our case. However, with a trick they can become perfectly nested.
+
+```cpp
+for(r=0;r<rows;r++){
+  for(c=0;c<cols;c++){
+      dot[c] = 0.0;
+      sum[c] = 0.0;
+      for(cc=(-2);cc<=2;cc++){
+        if(((c+cc) >= 0) && ((c+cc) < cols)){
+            dot[c] += (float)image[r*cols+(c+cc)] * kernel[center+cc];
+            sum[c] += kernel[center+cc];
+        }
+      }
+      tempim[r*cols+c] = dot[c]/sum[c];
+  }
+}
+```
+
+We converted temporary values `dot` and `sum` into arrays. By doing this, their values are preserved accross loop iterations. This allows us to break the loop over `c` into two loops:
+
+```cpp
+for(r=0;r<rows;r++){
+  for(c=0;c<cols;c++){
+      dot[c] = 0.0;
+      sum[c] = 0.0;
+  }
+  for(c=0;c<cols;c++){
+      for(cc=(-2);cc<=2;cc++){
+        if(((c+cc) >= 0) && ((c+cc) < cols)){
+            dot[c] += (float)image[r*cols+(c+cc)] * kernel[center+cc];
+            sum[c] += kernel[center+cc];
+        }
+      }
+  }
+  for(c=0;c<cols;c++){
+      tempim[r*cols+c] = dot[c]/sum[c];
+  }
+}
+```
+
+Loop over `c` and loop over `cc` are now perfectly nested and they can be interchanged.
```

Comment:
I suggest adding a link to where people can take a look at the final version of the code. Also, it would be great if we post a speedup for this single change.

Reply:
Need tome to do this.

Reply:
What is the "final version of the code"?

Reply:
We can go without a measured speedup for every code change or we can do it later.

Reply:
> What is the "final version of the code"?

I meant just to put a link(s) for submissions in which people can see transformed code.